### DISCUSSION TITLE
Change Str::random() to return Stringable

### DISCUSSION THREAD

**Discussion by PHLAK** [laravel/framework#35067] [Ideas]
Proposal
----------

Change the `Str::random()` method to return an instance of `Stringable` instead of a string to allow fluent chaining.

#### Example

```php
Str::random()->lower();
```

### Backwards Compatibility

Since the `Stringable` class implements the magic `__toString()` method this should have minimal impact to existing code. However, there is a chance for breakage if type hinting or comparing against a type of `string`.

The following are some of the situations I predict would cause issues.

```php
$randomString = Str::random();

if (gettype($randomString) === 'string') { ... } // gettype would return 'object'
```

```php
public function foo(string $string) { ... } // Expects a literal string

foo(Str::random()); // Would not work
```

However updating code could be as simple as declaring a type on creation.

```php
$randomString = (string) Str::random();
```

Background
-------------

Recently I had the need to generate random strings in the format of `xxxx-xxxx-xxxx-xxxx` (all lowercase) and came up with this

```php
implode('-', str_split(Str::lower(Str::random()), 4));
```

This works fine but it's a mix of Laravel `Str` helpers and native PHP methods. So I tried to use fluent strings to make this more expressive and readable.

```php
Str::random()->lower()->split(4)->implode('-');
```

This had two issues.

First, the `split()` method doesn't work the same as `str_split()`. For this I opened a [PR](https://github.com/laravel/framework/pull/35038) to add the `chunk()` method to the `Stringable` class. That was merged today.

The second issue is the reason for this idea. That is, the `Str::random()` method is not chainable.

If this idea is accepted and merged the following would become possible.

```php
Str::random()->lower()->chunk(4)->implode('-');
```

If there is interest I will do the work and open a PR.

**Comment by shaedrich** [comment]
Alternatively, they just add a shortcut to `Str::random` to `Stringable::random` like this:

```php
use Illuminate\Support\Stringable;

Stringable::macro('random', fn($length = 16) => tap($this, ($stringable) => $stringable->value = Str::random($length));

(new Stringable)->random()->lower();
```

**Comment by eapacheco** [comment]
I'm not convinced on stringable-first policy, as my little usual string handling can more easily be solved with the string helpers. There is another unmentioned solution already available, although not as elegant, it is definitely short:

```php
str(Str::random())->lower();
```

The result shaedrich achieved is somewhat elegant, like: `str()->random()` . But I'm concerned about the unexpected `append` behavior; and the extra difficulty it adds to linters.

**Reply by shaedrich** [reply]
Nice version 👍🏻 

I'm open for suggestions if you have a better way to write that macro :)

**Comment by rossjcooper** [comment]
I have opened a PR https://github.com/laravel/framework/pull/48612 to add the `random()` method to the Stringable class
